### PR TITLE
Pass and code meeting query tests

### PR DESCRIPTION
**Pull request by ninaprabhu**

### CODE REVIEW

**Comment by Tetheta** [comment]
Starting review

**Comment by Tetheta** [comment]
Ah.

Forgot to tell you. Please add the /node_modules directory to your .gitignore.

Otherwise you end up with a PR like this -> 700 

If yo can dot hat then I should be able to actually find your code in the 90000 lines 😆

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 1]
```diff
@@ -270,5 +271,120 @@ public void notEnoughRoom() {
 
     Assert.assertEquals(expected, actual);
   }
+
+  @Test
+  public void everyAttendeeIsConsideredOptionalNoChange() {
```
Comment (Tetheta) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:276]: If you can rework the names as we mentioned in standup that'll make it a bit easier to review. That way I know exactly what I'm looking for instead of having to figure out what NoChange means in this case from the code/etc.

File: portfolio/src/main/webapp/script.js [hunk 2]
```diff
@@ -102,4 +102,50 @@ function fetchBlobstoreUrl() {
...
+      // If the color-thief-output div is not in the viewport or cut off, scroll down.
+      var windowHeight          = $(window).height();
+      var currentScrollPosition = $('html').scrollTop()
+      var outputOffsetTop       = $imageSection.find('.color-thief-output').offset().top
+
+      if ((currentScrollPosition < outputOffsetTop) && (currentScrollPosition + windowHeight - 250 < outputOffsetTop)) {
+         $('html, body').animate({scrollTop: outputOffsetTop - windowHeight + 200 + "px"});
+      }
+    }, 300);
+  };
```
[the remark this answers is not in the document]
Comment (ninaprabhu) [reply portfolio/src/main/webapp/script.js:151]: oh yeah sorry about that! meant to change it later to something I understood better/worked for my specific website but didn't get around to it because I was running into issues like I mentioned before

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -13,11 +13,63 @@
...
+        TimeRange second = badTimes.get(i);
+        if (!first.overlaps(second) && (second.start() - first.end()) >= request.getDuration()){
+            goodTimes.add(TimeRange.fromStartEnd(first.end(), second.start(), false));
+        }
+        if (first.contains(second)){ // Replace second with first to extend end time and avoid later mistakes.
+            badTimes.set(i, TimeRange.fromStartDuration(first.start(), first.duration()));
+        }
+    }
+    int end = badTimes.get(badTimes.size()-1).end();
+    if (TimeRange.END_OF_DAY + 1 > end) goodTimes.add(TimeRange.fromStartEnd(end, TimeRange.END_OF_DAY+1, false));
```
[the remark this answers is not in the document]
Comment (ninaprabhu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:73]: To clarify, do we want spaces around the + or not? In the previous comment, there was an issue when there were no spaces around the -. The style guide says there should be spaces, so am I missing something for this comment?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -13,11 +13,63 @@
...
+    for (int i = 1; i < badTimes.size(); i++){
+        TimeRange first = badTimes.get(i-1);
+        TimeRange second = badTimes.get(i);
+        if (!first.overlaps(second) && (second.start() - first.end()) >= request.getDuration()){
+            goodTimes.add(TimeRange.fromStartEnd(first.end(), second.start(), false));
+        }
+        if (first.contains(second)){ // Replace second with first to extend end time and avoid later mistakes.
+            badTimes.set(i, TimeRange.fromStartDuration(first.start(), first.duration()));
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ninaprabhu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:71]: That makes sense - I wanted the change to be reflected when going to the next index and looking at i - 1. Would you suggest making an array copy and changing that? Or should I stick with this for now but try something different if I were to do this or something similar on my portfolio?

[the remark this answers is not in the document]
Comment (Tetheta) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:71]: Often for clarity in this kind of algorithm you'd use two passes (aka two for loops).
The first to consolidate the array into a new one without completely overlapping segments, then the second to go through and calculate the gaps/etc.

This has the disadvantage of slightly more iteration time and space (but still the same big O), but it has the significant advantage of additional clarity when coming back later and modifying things. Typically in our code we prefer the latter.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Tetheta) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:73]: I'm talking about the second + in the line here, sorry should've clarified that. We always want space around operators like that.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -184,7 +184,7 @@ private ImagePair getUploadedFileUrl(HttpServletRequest request, String formInpu
     // path to the image, rather than the path returned by imagesService which contains a host.
     try {
       URL url = new URL(imagesService.getServingUrl(options));
-      return new ImagePair(url.getPath(), imageLabel);
+      return new ImagePair(url.getPath(), imageLabel); // Getting an "cannot find symbol" error here.
     } catch (MalformedURLException e) {
       return new ImagePair(imagesService.getServingUrl(options), imageLabel);
```
Comment (ninaprabhu) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:189]: I'm getting a "symbol not found error" for imageLabel in both of these situations. I thought it was because of my try/catch statement earlier, but I return if there's an error, so I don't see how it would impact this.

Comment (LukeGMTravis) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:189]: I'm not sure what you mean by "in both of these situations". And "symbol not found" is all it says? Posting a screenshot of the error may be helpful.

As for the try/catch, IIRC in Java only the errors of the type noted in the catch statement are caught (or descendant types). You can always add more catch statements though to handle different kinds of (expected) errors.

Comment (Tetheta) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:189]: +1 to Luke's request for a screenshot.
Also what have you tried from googling the error/etc?
A quick google search on my part shows that typically it's when a class isn't imported. So I'd suspect the ImagePair class is funky somehow.

Comment (ninaprabhu) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:189]: I made the ImagePair class and it's in that file itself. When I looked it up on Google it said that it was usually a class not being imported or an uninitialized variable being called, so I'm not sure how that relates here. (Also sorry, I meant in both instances of the imagePair class being called.) Here's the whole error message:

Comment (ninaprabhu) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:189]: [ERROR] /home/ninaprabhu/google-2020/portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:[188,43] cannot find symbol
[ERROR]   symbol:   variable imageLabel
[ERROR]   location: class com.google.sps.servlets.DataServlet

Comment (ninaprabhu) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:189]: This happens on lines 187 and 189 - I think I moved some stuff around before pushing which is why it's showing up on line 188 in what I just pasted

Comment (Tetheta) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:189]: Have you tried making ImagePair its own file and importing it? I'm wondering if Java is getting confused because of the placement of it somehow. I think best practice is to have one class per file long term, so might want to adapt that here anyways.

Sorry for the late response on this.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -13,11 +13,71 @@
...
+  private ArrayList<TimeRange> makeBadTimes(Collection<Event> events, MeetingRequest request) {
+    ArrayList<TimeRange> badTimes = new ArrayList<>();
+    for (Event event : events) {
+      if (!Collections.disjoint(event.getAttendees(), request.getAttendees())) {
+        badTimes.add(event.getWhen());
+      }
+    }
+    Collections.sort(badTimes, TimeRange.ORDER_BY_START);
+    return badTimes;
+    }
```
Comment (Tetheta) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:54]: Nit: This should be indented two spaces less

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -13,11 +13,71 @@
...
+    ArrayList<TimeRange> badTimes = makeBadTimes(events, request);
+    makeGoodTimes(request, badTimes, goodTimes);
+    return goodTimes;
+  }
+
+  /* Returns a collection of TimeRanges during which at least one person in the request is busy. */
+  private ArrayList<TimeRange> makeBadTimes(Collection<Event> events, MeetingRequest request) {
+    ArrayList<TimeRange> badTimes = new ArrayList<>();
+    for (Event event : events) {
+      if (!Collections.disjoint(event.getAttendees(), request.getAttendees())) {
```
Comment (Tetheta) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:48]: This is a great use of the disjoint API on Collections.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -13,11 +13,71 @@
...
+        badTimes.add(event.getWhen());
+      }
+    }
+    Collections.sort(badTimes, TimeRange.ORDER_BY_START);
+    return badTimes;
+    }
+  
+  /* Returns a collection of TimeRanges that is the opposite of those specified in badTimes. */
+  private void makeGoodTimes(MeetingRequest request, ArrayList<TimeRange> badTimes, ArrayList<TimeRange> goodTimes) {
```
Comment (Tetheta) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:57]: Since you don't populate goodTimes with anything before calling this function, I think it'd be more ideal to return goodTimes here instead of mutating the passed in ArrayList.

This would give greater clarity to your initial query function and clean up the return statement you're forced to use in here a bit.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -13,11 +13,71 @@
...
+        return result; //We found times that worked or there are no mandatory attendees.
+    }
+    return queryHelper(events, request); 
+  }
+
+  /* Returns a collection of TimeRanges that satisfy the people and duration needed in the request. */
+  private Collection<TimeRange> queryHelper(Collection<Event> events, MeetingRequest request) {
+    ArrayList<TimeRange> goodTimes = new ArrayList<>();
+    if (request.getDuration() > TimeRange.END_OF_DAY) {
+        return goodTimes;
```
Comment (Tetheta) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:37]: Referencing my comment on makeGoodTimes below, you could then remove the initialization here and just return an empty arrayList directly in the if statement.

It also then matches makeBadTimes's API which makes more intuitive sense.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -13,11 +13,71 @@
...
+  
+  /* Returns a collection of TimeRanges that is the opposite of those specified in badTimes. */
+  private void makeGoodTimes(MeetingRequest request, ArrayList<TimeRange> badTimes, ArrayList<TimeRange> goodTimes) {
+    if (badTimes.size() == 0) { // We are free the whole day - no times are off limits.
+      goodTimes.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY + 1, false));
+      return;
+    }
+
+    int duration = badTimes.get(0).start();
+    if (duration > 0) {
```
Comment (Tetheta) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:64]: Duration isn't used again so it's only the first event duration here right? I'd rename it as such just to be clear it's not going to be used again later.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -13,11 +13,71 @@
...
+        goodTimes.add(TimeRange.fromStartDuration(TimeRange.START_OF_DAY, duration));
+    }
+    
+    for (int i = 1; i < badTimes.size(); i++){
+        TimeRange first = badTimes.get(i - 1);
+        TimeRange second = badTimes.get(i);
+        if (!first.overlaps(second) && (second.start() - first.end()) >= request.getDuration()) {
```
Comment (Tetheta) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:71]: An explanatory comment could be useful here so it's easier to understand at a glance.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 12]
```diff
@@ -270,5 +271,120 @@ public void notEnoughRoom() {
...
+        new Event("Event 1", TimeRange.fromStartDuration(TIME_0800AM, DURATION_30_MINUTES),
+            Arrays.asList(PERSON_A)),
+        new Event("Event 2", TimeRange.fromStartDuration(TIME_0900AM, DURATION_30_MINUTES),
+            Arrays.asList(PERSON_B)),
+        new Event("Event 3", TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true),
+            Arrays.asList(PERSON_C)));
+
+    MeetingRequest request =
+        new MeetingRequest(Arrays.asList(PERSON_A, PERSON_B), DURATION_30_MINUTES);
+        request.addOptionalAttendee(PERSON_C);
```
Comment (Tetheta) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:287]: If this is called "everyAttendeeIsConsideredOptional" why is C the only optional attendee?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 5 above)
Comment (ninaprabhu) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: The issue seemed to be with defining imageLabel in a try/catch statement. I thought that if it passed through successfully imageLabel would be defined but apparently it only exists within the scope of the try/catch statement? I ended up initializing it beforehand and everything worked fine! I also moved ImagePair to its own file for better style after :)

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 13]
```diff
@@ -270,5 +271,120 @@ public void notEnoughRoom() {
...
+        new Event("Event 1", TimeRange.fromStartDuration(TIME_0800AM, DURATION_30_MINUTES),
+            Arrays.asList(PERSON_A)),
+        new Event("Event 2", TimeRange.fromStartDuration(TIME_0900AM, DURATION_30_MINUTES),
+            Arrays.asList(PERSON_B)),
+        new Event("Event 3", TimeRange.fromStartDuration(TIME_0830AM, DURATION_30_MINUTES),
+            Arrays.asList(PERSON_C)));
+
+    MeetingRequest request =
+        new MeetingRequest(Arrays.asList(PERSON_A, PERSON_B), DURATION_30_MINUTES);
+        request.addOptionalAttendee(PERSON_C);
```
Comment (Tetheta) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:310]: Same question on this one. Might be I'm misunderstanding something

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 14]
```diff
@@ -270,5 +271,120 @@ public void notEnoughRoom() {
...
+  }
+
+  @Test
+  public void everyAttendeeIsConsideredAllOptional() {
+    // Have each person have different events. We should see two options because each person has
+    // split the restricted times.
+    //
+    // Events  :       |--A--|     |--B--|
+    // Day     : |-----------------------------|
+    // Options : |--1--|     |--2--|     |--3--|
```
Comment (Tetheta) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:347]: These little diagrams are excellent, I'd recommend adding one on each test.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 15]
```diff
@@ -270,5 +271,120 @@ public void notEnoughRoom() {
...
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartDuration(TIME_0830AM, DURATION_30_MINUTES));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void everyAttendeeIsConsideredAllOptional() {
```
Comment (Tetheta) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:341]: I think I'm not understanding the meaning of your test names.

Can you clarify? Sometimes it's helpful to use underscores if you're trying to separate setup from expected outcomes.
Or use a keyword "Should" or something to separate.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 5 above)
Comment (Tetheta) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Ah didn't realize that was happening, should've looked at the line in context, this little comment view only shows what's below it lol. 

The way Java works is that context visibility is determined by {} brackets. So if something is declared inside, only other code inside can see it.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 12 above)
Comment (ninaprabhu) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:287]: Sorry, I meant that optional attendees are present. I'll modify the name of the test to make this more clear.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 8 above)
Comment (ninaprabhu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:57]: That makes sense! I think I just didn't do this earlier for the one edge case when the duration of the request was more than a day, but I just changed that to return an empty unnamed list and incorporated your feedback

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (ninaprabhu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:48]: Thank you! :D

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ninaprabhu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: For some reason, I totally did not realize that implementing this is super simple until today - just did this!

File: portfolio/src/main/webapp/script.js [hunk 16]
```diff
@@ -12,23 +12,23 @@
...
 }
 
-/* Display comments */
+/* Display comments. */
 function getComments() {
     document.getElementById('show-comments').innerHTML = "";
     let num = document.getElementById("num-show").value;
     num = parseInt(num);
-    if (!(isNaN(num) || num == 0)) {
+    if (!(isNaN(num) || num==0)) {
```
Comment (Tetheta) [inline portfolio/src/main/webapp/script.js:31]: Is removing the spaces something prettier did? Typically we have spaces around ==

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 17]
```diff
@@ -51,33 +50,44 @@
     }
     Collections.sort(badTimes, TimeRange.ORDER_BY_START);
     return badTimes;
-    }
+  }
   
   /* Returns a collection of TimeRanges that is the opposite of those specified in badTimes. */
-  private void makeGoodTimes(MeetingRequest request, ArrayList<TimeRange> badTimes, ArrayList<TimeRange> goodTimes) {
+  private ArrayList<TimeRange> makeGoodTimes(MeetingRequest request, ArrayList<TimeRange> badTimes) {
+    ArrayList<TimeRange> badTimesCopy = badTimes;
```
Comment (Tetheta) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:57]: This doesn't actually make a copy, it just makes another variable pointing at the badTimes ArrayList. So while it works in your case since you want to modify the original array, it's going to cause problems down the road.

Use the .clone() method on the badTimes ArrayList instead, I think that should do the trick.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 8 above)
Comment (Tetheta) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Looks great :)

File: portfolio/src/main/webapp/script.js (hunk 16 above)
Comment (ninaprabhu) [reply portfolio/src/main/webapp/script.js:31]: It might have been - I definitely had the spaces around it in an earlier version. Maybe I was messing around with it before prettier was implemented and it didn't change it?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 17 above)
Comment (ninaprabhu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:57]: Ooh makes sense - didn't realize that happens in Java as well. Will implement that before merging!

Comment (ninaprabhu) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:57]: Just wanted to note here that I added a while loop to account for edge cases (not something that shows up in the tests as far as I know, but still could occur.) Also, now that we're not modifying the array in place, I can use the remove keyword instead of setting two array indices to the same thing.
